[expo-application] create expo-application unimodule

ninjaachibi opened the pull request:
# Why

Following the `expo-device` and `expo-battery` unimodules, we created the `application` unimodule to expose constants and methods related to the application on the device, such as `bundleId`, `firstInstallTime`, and `nativeAppVersion`.

# How

We moved some the functions we had originally in the `expo-device` and `expo-constants` modules that made more sense in being in a separate `application` module. We also referenced some of the implementations in [RN-device-info](https://github.com/react-native-community/react-native-device-info). 

# Test Plan

We tested in sandbox on Android and iOS devices.

ninjaachibi committed:
create expo-application unimodule

Co-authored-by: Robert Luo <1998.robert@gmail.com>
Co-authored-by: Vivian Zhu <tongyu@usc.edu>

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,130 @@
+
+### `Application.getIosIdForVendorAsync()`
+
+**iOS only.** Gets the iOS identifier for vender [(IDFV)](https://developer.apple.com/documentation/uikit/uidevice/1620059-identifierforvendor) value, an alphanumeric string that uniquely identifies a device to the app’s vendor.
```
```suggestion
**iOS only.** Gets the iOS identifier for vendor [(IDFV)](https://developer.apple.com/documentation/uikit/uidevice/1620059-identifierforvendor) value, an alphanumeric string that uniquely identifies a device to the app’s vendor.
```

Also need to document that, when, and why this could be null.

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,130 @@
+
+**iOS only.** Gets the iOS identifier for vender [(IDFV)](https://developer.apple.com/documentation/uikit/uidevice/1620059-identifierforvendor) value, an alphanumeric string that uniquely identifies a device to the app’s vendor.
+
+Since this is the IDFV value, its value can change if all apps from the current apps vendor have been uninstalled.
```
```suggestion
The OS will change the vendor identifier if all apps from the current apps' vendor have been uninstalled.
```

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,130 @@
+
+### `Application.applicationName`
+
+The human-readable name of the application.
```
Need to be comprehensive -- the easy stuff is easy to guess anyway so most of the value in the docs is the stuff that's hard to find out.

```suggestion
The human-readable name of the application that is displayed with the app's icon on the device's home screen or desktop. On Android and iOS, this value is a string unless the name could not be retrieved, in which case this value will be null. On web this value is null.
```

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,130 @@
+
+### `Application.bundleId`
+
+The bundle ID of the application.
```
```suggestion
The ID of the application. On Android, this is the application ID. On iOS this is the bundle ID. On web, this is null.
```

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,130 @@
+
+### `Application.nativeAppVersion`
+
+The `Info.plist` value for `CFBundleShortVersionString` on iOS and the version name set by `version` in app.json on Android at the time the native app was built.
```
```suggestion
The human-readable version of the native application that may be displayed in the app store. This is the `Info.plist` value for `CFBundleShortVersionString` on iOS and the version name set by `version` in app.json on Android at the time the native app was built. On web, this value is null.
```

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,130 @@
+
+### `Application.nativeBuildVersion`
+
+The `Info.plist` value for `CFBundleVersion` on iOS (set with `ios.buildNumber` value in `app.json` in a standalone app) and the version code set by `android.versionCode` in app.json on Android at the time the native app was built.
```
```suggestion
The internal build version of the native application that the app store may use to distinguish between different binaries. This is the `Info.plist` value for `CFBundleVersion` on iOS (set with `ios.buildNumber` value in `app.json` in a standalone app) and the version code set by `android.versionCode` in app.json on Android at the time the native app was built. On web, this value is null.
```

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,130 @@
+
+The `Info.plist` value for `CFBundleVersion` on iOS (set with `ios.buildNumber` value in `app.json` in a standalone app) and the version code set by `android.versionCode` in app.json on Android at the time the native app was built.
+
+- e.g., `2.11.0`
```
```suggestion
- e.g., `2.11.0.16344`
```

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,130 @@
+
+### `Application.androidId`
+
+**Android only.** A hexademical `string` unique to each combination of app-signing key, user, and device. The value may change if a factory reset is performed on the device or if an APK signing key changes. For more information about how the platform handles ANDROID_ID in Android 8.0 (API level 26) and higher, see [Android 8.0 Behavior Changes.](https://developer.android.com/about/versions/oreo/a ...
```
```suggestion
**Android only.** The value of [`Settings.Secure.ANDROID_ID`](https://developer.android.com/reference/android/provider/Settings.Secure.html#ANDROID_ID). This is a hexademical `string` unique to each combination of app-signing key, user, and device. The value may change if a factory reset is performed on the device or if an APK signing key changes. For more information about how the platform handles `ANDROID_ID` in Android 8.0 (API level 26) and higher, see [Android 8.0 Behavior Changes.](https://developer.android.com/about/versions/oreo/android-8.0-changes.html#privacy-all)
```

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,130 @@
+
+Note: In versions of the platform lower than Android 8.0 (API level 26), this value remains constant for the lifetime of the user's device.
+
+See: https://developer.android.com/reference/android/provider/Settings.Secure.html#ANDROID_ID
```
Remove this after linking it above.

ide commented on docs/pages/versions/unversioned/sdk/application.md:
````diff
@@ -0,0 +1,130 @@
+
+```js
+await Application.getIosIdForVendorAsync();
+//"FCDBD8EF-62FC-4ECB-B2F5-92C9E79AC7F9"
````
```suggestion
// "FCDBD8EF-62FC-4ECB-B2F5-92C9E79AC7F9"
```

ide commented on docs/pages/versions/unversioned/sdk/application.md:
````diff
@@ -0,0 +1,130 @@
+
+```js
+await Application.getInstallReferrerAsync();
+//"utm_source=google-play&utm_medium=organic"
````
```suggestion
// "utm_source=google-play&utm_medium=organic"
```

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,130 @@
+
+### `Application.getInstallReferrerAsync()`
+
+**Android only.** Gets the referrer URL of the installed app with [`Install Referrer API`](https://developer.android.com/google/play/installreferrer) from Google Play Store.
```
```suggestion
**Android only.** Gets the referrer URL of the installed app with [`Install Referrer API`](https://developer.android.com/google/play/installreferrer) from the Google Play Store.
```

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,130 @@
+
+### `Application.getFirstInstallTimeAsync()`
+
+Gets the first time the app was installed onto the device.
```
```suggestion
Gets the time the app was first ever installed onto the device.
```

(is this accurate? and how does it work with multiple user accounts?) What about web?

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,130 @@
+
+### `Application.getLastUpdateTimeAsync()`
+
+**Android only.** Gets the last time the app was updated from the App store or Play Store.
```
```suggestion
**Android only.** Gets the last time the app was updated from the Google Play Store.
```
(this is android-only)

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,130 @@
+
+#### Returns
+
+Returns a `Promise` that resolves to a `Date` object that represents the last time the app was updated via the App Store (iOS) or Play Store (Android).
```
```suggestion
Returns a `Promise` that resolves to a `Date` object that represents the last time the app was updated via the Google Play Store (Android).
```

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,130 @@
+title: Application
+---
+
+This module provides useful information about the native application, such the bundle ID, app name, and build version.
```
```suggestion
This module provides useful information about the native application, such its ID, app name, and build version.
```

ide commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,159 @@
+      PackageInfo pInfo = packageManager.getPackageInfo(packageName, 0);
+      constants.put("nativeAppVersion", pInfo.versionName);
+      constants.put("nativeBuildVersion", pInfo.versionCode);
+    } catch (PackageManager.NameNotFoundException e) {
```
Under what conditions can this happen? We should document that..

ninjaachibi replied:
This error is thrown "when a given package, application, or component name cannot be found". I don't really expect this to fail since we grab the package name from `mContext`, so we should expect it to be the correct `bundleId` of the package. However, `getPackageInfo()` throws this error, so we have to catch it in a `catch` block.

ide commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,159 @@
+      promise.resolve((double)info.lastUpdateTime);
+    } catch (Exception e) {
+      e.printStackTrace();
+      promise.reject("ERR_APPLICATION", "Unable to get last update time of this application.", e);
```
This needs to be documented (see screen orientation's docs for how to document error codes), it is part of the API.

ide commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,159 @@
+
+  @ExpoMethod
+  public void getInstallReferrerAsync(final Promise promise) {
+
```
```suggestion
```

ide commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,159 @@
+              referrerClient.endConnection();
+            } catch (RemoteException e) {
+              e.printStackTrace();
+              promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "Binding to play store remote invocation error.", e);
```
This error message is really cryptic.
```suggestion
              promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "The Play Store raised an error when retrieving the install referrer", e);
```

ide commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,159 @@
+            referrerClient.endConnection();
+            break;
+          default:
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "Connection could not be established");
```
```suggestion
            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "Could not establish a connection to Google Play");
```

ide commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,159 @@
+      @Override
+      public void onInstallReferrerServiceDisconnected() {
+        // Try to restart the connection on the next request to
+        // Google Play by calling the startConnection() method.
```
This seems dangerous if it can cause the JS code to hang waiting for a promise

ninjaachibi replied:
`void onInstallReferrerServiceDisconnected();` is declared in the interface for the `InstallReferrerStateListener` class, so what would you suggest we do?

esamelson replied:
might make sense to save the promise as a member of the class and reject it here

esamelson replied:
this still needs to be resolved, i think -- we definitely want to avoid a hanging promise that never resolves. we may want to keep track of whether or not the promise has been resolved/rejected, and reject it here if it hasn't been

ide commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,159 @@
+          case InstallReferrerClient.InstallReferrerResponse.FEATURE_NOT_SUPPORTED:
+            // API not available on the current Play Store app
+            Log.d("INSTALL_REFERRER", "feature not supported");
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "This feature not supported");
```
This too:
```suggestion
            promise.reject("ERR_APPLICATION_INSTALL_REFERRER_UNAVAILABLE", "This app doesn't provide the installation referrer");
```

ide commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,159 @@
+              installReferrer.append(response.getInstallReferrer());
+              response.getReferrerClickTimestampSeconds();
+              response.getInstallBeginTimestampSeconds();
+              promise.resolve(installReferrer.toString());
```
Move this outside the try-catch; we don't expect it to fail

ide commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,159 @@
+            break;
+          default:
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "Connection could not be established");
+            referrerClient.endConnection();
```
Seems like we always want to end the connection -- better to do it in a finally block

ide commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,159 @@
+          case InstallReferrerClient.InstallReferrerResponse.SERVICE_UNAVAILABLE:
+            // Connection could not be established
+            Log.d("INSTALL_REFERRER", "connection could not be established");
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "Connection could not be established");
```
```suggestion
            promise.reject("ERR_APPLICATION_INSTALL_REFERRER_CONNECTION", "Could not establish a connection to Google Play");
```

ide commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,159 @@
+            }
+            break;
+          case InstallReferrerClient.InstallReferrerResponse.FEATURE_NOT_SUPPORTED:
+            // API not available on the current Play Store app
```
```suggestion
            // API not available in the current Play Store app
```

ide commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,159 @@
+      public void onInstallReferrerSetupFinished(int responseCode) {
+        switch (responseCode) {
+          case InstallReferrerClient.InstallReferrerResponse.OK:
+            // Connection established
```
```suggestion
            // Connection established and response received
```

ide commented on packages/expo-application/ios/EXApplication/EXApplication.m:
```diff
@@ -0,0 +1,74 @@
+}
+
+- (NSString *)appleIFV {
+  if(NSClassFromString(@"UIDevice") && [UIDevice instancesRespondToSelector:@selector(identifierForVendor)]) {
```
Don't need this, just call identifierForVendor always (remove the extra method)

ide commented on packages/expo-application/ios/EXApplication/EXApplication.m:
```diff
@@ -0,0 +1,74 @@
+
+UM_EXPORT_METHOD_AS(getIosIdForVendorAsync, getIosIdForVendorAsyncWithResolver:(UMPromiseResolveBlock)resolve rejecter:(UMPromiseRejectBlock)reject)
+{
+  NSString* identifierForVendor = [self appleIFV];
```
handle nil -> NSNull

ide commented on packages/expo-application/ios/EXApplication/EXApplication.m:
```diff
@@ -0,0 +1,74 @@
+  return nil;
+}
+
+- (NSString *) applicationName
```
just inline all these methods, easy enough to understand without the extra lines of code

ide commented on packages/expo-application/ios/EXApplication/EXApplication.m:
```diff
@@ -0,0 +1,74 @@
+UM_EXPORT_METHOD_AS(getFirstInstallTimeAsync, getFirstInstallTimeAsyncWithResolver:(UMPromiseResolveBlock)resolve rejecter:(UMPromiseRejectBlock)reject)
+{
+  NSURL* urlToDocumentsFolder = [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] lastObject];
+  __autoreleasing NSError *error;
```
```suggestion
 NSError *error;
```

ide commented on packages/expo-application/ios/EXApplication/EXApplication.m:
```diff
@@ -0,0 +1,74 @@
+{
+  NSURL* urlToDocumentsFolder = [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] lastObject];
+  __autoreleasing NSError *error;
+  NSDate *installDate = [[[NSFileManager defaultManager] attributesOfItemAtPath:urlToDocumentsFolder.path error:&error] objectForKey:NSFileCreationDate];
```
Need to handle the error

ide commented on packages/expo-application/ios/EXApplication/EXApplication.m:
```diff
@@ -0,0 +1,74 @@
+{
+  NSURL* urlToDocumentsFolder = [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] lastObject];
+  __autoreleasing NSError *error;
+  NSDate *installDate = [[[NSFileManager defaultManager] attributesOfItemAtPath:urlToDocumentsFolder.path error:&error] objectForKey:NSFileCreationDate];
```
Is this the first install time or the last install time?

vivianzzhu91 replied:
It's the first install date because it's checking `NSFileCreationDate` of this document directory.

ide replied:
iOS will clear an app's data when you delete the app, though. Have you verified that it always preserves the ctime?

vivianzzhu91 replied:
We both checked on iOS and Android, and it seemed that uninstalling the app would delete all data, and would return the reinstalled date if reinstalled.

ide commented on packages/expo-application/ios/EXApplication/EXApplication.m:
```diff
@@ -0,0 +1,74 @@
+  NSURL* urlToDocumentsFolder = [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] lastObject];
+  __autoreleasing NSError *error;
+  NSDate *installDate = [[[NSFileManager defaultManager] attributesOfItemAtPath:urlToDocumentsFolder.path error:&error] objectForKey:NSFileCreationDate];
+  NSTimeInterval timeInMiliseconds = [installDate timeIntervalSince1970]*1000;
```
```suggestion
  NSTimeInterval timeInMiliseconds = [installDate timeIntervalSince1970] * 1000;
```

ide commented on packages/expo-application/src/Application.ts:
```diff
@@ -0,0 +1,45 @@
+import { Platform, UnavailabilityError } from '@unimodules/core';
```
run prettier over this file

ide commented on packages/expo-application/src/ExpoApplication.web.ts:
```diff
@@ -0,0 +1,5 @@
+export default {
```
prettier

ide commented on packages/expo-application/src/Application.ts:
```diff
@@ -0,0 +1,45 @@
+export const applicationName = ExpoApplication ? ExpoApplication.applicationName : null;
+export const bundleId = ExpoApplication ? ExpoApplication.bundleId : null;
+export let androidId;
+if(Platform.OS === 'ios'){
```
need to handle web

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,130 @@
+
+- e.g., `com.cocoacasts.scribbles`, `com.apple.Pages`
+
+### `Application.nativeAppVersion`
```
```suggestion
### `Application.nativeApplicationVersion`
```

(unabbreviate consistently)

vivianzzhu91 committed:
[docs] update docs description

vivianzzhu91 committed:
[web] handle web constants and update format

vivianzzhu91 committed:
[ios] cleanup inline methods and handle error with NSError

ninjaachibi committed:
[docs] add error codes

ninjaachibi committed:
[android] add error codes

* took out endConnection and things work
* add error codes

vivianzzhu91 committed:
[ios] resolve conditional logic and resolve promise twice

vivianzzhu91 committed:
[docs] update getFirstInstallTimeAsync changes description

ninjaachibi committed:
[android] close install referral connection

we didn't close the connection before, which could be dangerous/cause memory leaks. checked and this works

vivianzzhu91 committed:
[web] add nullable constants to web

ninjaachibi committed:
[docs/android] update error codes

* Add ERR_APPLICATION_PACKAGE_NAME_NOT_FOUND code
* Fix soem things in docs

vivianzzhu91 committed:
[ios] add conditionals to avoid double invocation of promise

ninjaachibi committed:
[docs] add note about types for nativeBuildVersion

it seems that in the native code, for iOS the nativeBuildVersion in Info.plist is a string, while android.versionCode is an int

vivianzzhu91 committed:
[docs] update installation guide in README and docs

esamelson commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,153 @@
+      PackageInfo info = packageManager.getPackageInfo(packageName, 0);
+      promise.resolve((double)info.firstInstallTime);
+    } catch (PackageManager.NameNotFoundException e) {
+      e.printStackTrace();
```
use `Log.e(TAG, e)` instead in all these places

esamelson commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,153 @@
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER_CONNECTION", "Could not establish a connection to Google Play");
+            break;
+          default:
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "General error");
```
let's include the `responseCode` in the message here

esamelson reviewed (changes requested):
this is looking good -- thanks for your thoroughness on the documentation.

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,142 @@
+
+### `Application.bundleId`
+
+The ID of the application. On Android, this is the application ID. On iOS, this is the bundle ID. On web, this is `null`.
```
Android calls this the application ID: https://developer.android.com/studio/build/application-id

esamelson replied:
oh, oops -- ignore my previous comment as well then

ide reviewed (changes requested):
It sounds like across Android and iOS, "first install time" is the time that the app was most recently installed, not counting updates. If this is accurate, we should rename the method to just "installation time" and explain the behavior clearly in the docs.

vivianzzhu91 committed:
rename getFirstInstallTimeAsync to getInstallationTimeAsync and bundleId to applicationId

ninjaachibi committed:
[android] change constant and method names

* bundleId->applicationId
* getFirstInstallTimeAsync() -> getInstallationTimeAsync()

ninjaachibi committed:
[android/docs] add debugger log statements

ninjaachibi committed:
Merge branch 'master' into @expo-application

vivianzzhu91 committed:
update docs and code format in js

esamelson commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,153 @@
+        switch (responseCode) {
+          case InstallReferrerClient.InstallReferrerResponse.OK:
+            // Connection established and response received
+            Log.d("INSTALL_REFERRER", "connection established and response ok");
```
remove all the `Log.d`s

esamelson reviewed (changes requested):
this looks good, just still need to resolve the one comment from before

ninjaachibi committed:
Merge branch 'master' into @expo-application

ninjaachibi committed:
[ios] update pods

ninjaachibi committed:
[android/docs] add promise rejection for hanging promise

* optimize imports
* add promise rejection + error code
* remove Log.d

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,141 @@
+
+### `Application.getInstallationTimeAsync()`
+
+Gets the time the app was first installed onto the device. The time would change to the reinstalled date if uninstall and reinstall the app.
```
```suggestion
Gets the time the app was installed onto the device, not counting subsequent updates. If the app is uninstalled and reinstalled, this method returns the time the app was reinstalled.
```

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,141 @@
+
+| Code                                              | Description                                                                                                                                                                                                                                                                                                                                               ...
+| ------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
+| ERR_APPLICATION_PACKAGE_NAME_NOT_FOUND            | Error code thrown for `firstInstallTimeAsync` and `lastUpdateTimeAsync`. This may be thrown if the package information or package name could not be retrieved.                                                                                                                                                                                            ...
```
```suggestion
| ERR_APPLICATION_PACKAGE_NAME_NOT_FOUND            | Error code thrown by `getInstallationTimeAsync ` and `getLastUpdateTimeAsync`. This may be thrown if the package information or package name could not be retrieved.                                                                                                                                                                                                                                |
```

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,141 @@
+| ERR_APPLICATION_INSTALL_REFERRER_UNAVAILABLE      | The current Play Store app doesn't provide the installation referrer API, or the Play Store may not be installed. This error code may come up when testing on an AVD that doesn't come with the Play Store pre-installed, such as the Google Pixel 3 and Nexus 6.                                                                                         ...
+| ERR_APPLICATION_INSTALL_REFERRER_CONNECTION       | A connection could not be established to the Google Play Store                                                                                                                                                                                                                                                                                            ...
+| ERR_APPLICATION_INSTALL_REFERRER_REMOTE_EXCEPTION | A `RemoteException` was thrown after a connection was established to the Play Store. This may happen if the process hosting the remote object is no longer available, which usually means the process crashed. See https://stackoverflow.com/questions/3156389/android-remoteexceptions-and-services.                                                     ...
+| ERR_APPLICATION_INSTALL_REFERRER                  | General default case error code for the `getInstallReferrerAsync` method. This error code will be thrown if an exception occured for the install referrer, but the exception was none of the errors above. The [`responseCode`](https://developer.android.com/reference/com/android/installreferrer/api/InstallReferrerClient.InstallReferrerResponse.htm ...
```
```suggestion
| ERR_APPLICATION_INSTALL_REFERRER                  | General default case error code for the `getInstallReferrerAsync` method. This error code will be thrown if an exception occurred when getting the install referrer, but the exception was none of the more precise errors. The [`responseCode`](https://developer.android.com/reference/com/android/installreferrer/api/InstallReferrerClient.InstallReferrerResponse.html) is provided along with the error. |
```

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,141 @@
+| ERR_APPLICATION_INSTALL_REFERRER_CONNECTION       | A connection could not be established to the Google Play Store                                                                                                                                                                                                                                                                                            ...
+| ERR_APPLICATION_INSTALL_REFERRER_REMOTE_EXCEPTION | A `RemoteException` was thrown after a connection was established to the Play Store. This may happen if the process hosting the remote object is no longer available, which usually means the process crashed. See https://stackoverflow.com/questions/3156389/android-remoteexceptions-and-services.                                                     ...
+| ERR_APPLICATION_INSTALL_REFERRER                  | General default case error code for the `getInstallReferrerAsync` method. This error code will be thrown if an exception occured for the install referrer, but the exception was none of the errors above. The [`responseCode`](https://developer.android.com/reference/com/android/installreferrer/api/InstallReferrerClient.InstallReferrerResponse.htm ...
+| ERR_APPLICATION_INSTALL_REFERRER_SERVICE_DISCONNECTED | Connection to the install referrer service was lost. This error is thrown when an attempt was made to connect and set up the install referrer service, but the connection was lost. See the [Android documentation](https://developer.android.com/reference/com/android/installreferrer/api/InstallReferrerStateListener) for more information|
```
```suggestion
| ERR_APPLICATION_INSTALL_REFERRER_SERVICE_DISCONNECTED | Connection to the install referrer service was lost. This error is thrown when an attempt was made to connect and set up the install referrer service, but the connection was lost. See the [Android documentation](https://developer.android.com/reference/com/android/installreferrer/api/InstallReferrerStateListener) for more information. |
```

ide commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,148 @@
+
+      @Override
+      public void onInstallReferrerServiceDisconnected() {
+        // Try to restart the connection on the next request to
```
This comment doesn't look relevant anymore with the promise rejection -- delete?

ninjaachibi replied:
yep you're right. deleting

ide commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,148 @@
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER_CONNECTION", "Could not establish a connection to Google Play");
+            break;
+          default:
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "General error: response code " + responseCode);
```
Seeing "General error: response code XX" in your error logs without context is going to be hard to debug -- add more context:

```suggestion
            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "General error retrieving the install referrer: response code " + responseCode);
```

ide commented on packages/expo-application/ios/EXApplication/EXApplication.m:
```diff
@@ -0,0 +1,55 @@
+
+UM_EXPORT_METHOD_AS(getIosIdForVendorAsync, getIosIdForVendorAsyncWithResolver:(UMPromiseResolveBlock)resolve rejecter:(UMPromiseRejectBlock)reject)
+{
+  if(NSClassFromString(@"UIDevice") && [UIDevice instancesRespondToSelector:@selector(identifierForVendor)]) {
```
We probably don't need this check -- on all the devices we support `- [UIDevice identifierForVendor]` should work I think (or is that not the case)?

vivianzzhu91 replied:
It should work across all devices. I'll remove this check!

ide commented on packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java:
```diff
@@ -0,0 +1,148 @@
+    try {
+      PackageInfo pInfo = packageManager.getPackageInfo(packageName, 0);
+      constants.put("nativeApplicationVersion", pInfo.versionName);
+      constants.put("nativeBuildVersion", pInfo.versionCode);
```
versionCode is deprecated -- use getLongVersionCode() and get the lower 32 bits. then convert to a string so we are consistent across platforms.

ide commented on packages/expo-application/ios/EXApplication/EXApplication.m:
```diff
@@ -0,0 +1,55 @@
+#import <EXApplication/EXApplication.h>
+#import <UIKit/UIKit.h>
+
+@interface EXApplication ()
```
remove private interface if it isn't used

ide commented on docs/pages/versions/unversioned/sdk/application.md:
```diff
@@ -0,0 +1,141 @@
+
+### `Application.androidId`
+
+**Android only.** The value of [`Settings.Secure.ANDROID_ID`](https://developer.android.com/reference/android/provider/Settings.Secure.html#ANDROID_ID). This is a hexademical `string` unique to each combination of app-signing key, user, and device. The value may change if a factory reset is performed on the device or if an APK signing key changes. For more information about how the platform handl ...
```
```suggestion
**Android only.** The value of [`Settings.Secure.ANDROID_ID`](https://developer.android.com/reference/android/provider/Settings.Secure.html#ANDROID_ID). This is a hexademical `string` unique to each combination of app-signing key, user, and device. The value may change if a factory reset is performed on the device or if an APK signing key changes. For more information about how the platform handles `ANDROID_ID` in Android 8.0 (API level 26) and higher, see [Android 8.0 Behavior Changes.](https://developer.android.com/about/versions/oreo/android-8.0-changes.html#privacy-all) On iOS and web, this value is `null`.
```

ide commented on packages/expo-application/ios/EXApplication/EXApplication.m:
```diff
@@ -0,0 +1,55 @@
+  NSDate *installDate = [[[NSFileManager defaultManager] attributesOfItemAtPath:urlToDocumentsFolder.path error:&error] objectForKey:NSFileCreationDate];
+  if(error){
+    reject(@"ERR_APPLICATION", @"Unable to get installation time of this application.", error);
+  }
```
we use something closer to k&r, not Stroustrup

```suggestion
  } else {
```

ide commented on packages/expo-application/ios/EXApplication/EXApplication.m:
```diff
@@ -0,0 +1,55 @@
+  }
+  else{
+    NSTimeInterval timeInMiliseconds = [installDate timeIntervalSince1970] * 1000;
+    NSNumber* timeInterval = @(timeInMiliseconds);
```
c-style instead of c++ style, also "timeInterval" is a confusing name esp when this is not an NSTimeInterval
```suggestion
    NSNumber *timeNumber = @(timeInMilliseconds);
```

ide commented on packages/expo-application/ios/EXApplication/EXApplication.m:
```diff
@@ -0,0 +1,55 @@
+    reject(@"ERR_APPLICATION", @"Unable to get installation time of this application.", error);
+  }
+  else{
+    NSTimeInterval timeInMiliseconds = [installDate timeIntervalSince1970] * 1000;
```
```suggestion
    NSTimeInterval timeInMilliseconds = [installDate timeIntervalSince1970] * 1000;
```

vivianzzhu91 commented on docs/pages/versions/unversioned/sdk/application.md:
````diff
@@ -0,0 +1,141 @@
+
+```js
+await Application.getInstallReferrerAsync();
+// "utm_source=google-play&utm_medium=organic"
````
This is all the result that we log out from `installReferrer `.

vivianzzhu91 committed:
[docs] update docs format

vivianzzhu91 committed:
[web] update getFirstInstallationTime to return null on web

ninjaachibi committed:
Merge branch 'master' into @expo-application

ninjaachibi committed:
[ios] reinstall pods

ran pod install to include `expo-application`

vivianzzhu91 committed:
[ios] update c-style code format and remove checks

vivianzzhu91 committed:
cast nativeBuildVersion to be string

ninjaachibi committed:
Update build.gradle

Update compilesdk version to 28

ninjaachibi committed:
Update ApplicationModule.java

* add more context on general case error
* add current method to get version code for api version 28+

ide committed:
Update application.md

ide committed:
Update EXApplication.m

ide committed:
Update application.md

LinusU commented:
Sorry if this is not the right place to ask, couldn't find the information elsewhere 🙈

When is this module going to be published to npm? We are currently using `expo-constants` (using unimodules) but need access to the bundle id (`applicationId`) as well ☺️
